contracts/CreditorProxy.sol

```diff
@@ -0,0 +1,409 @@
+    address public TOKEN_TRANSFER_PROXY;
+    bytes32 constant public NULL_ISSUANCE_HASH = bytes32(0);
+    uint16 constant public EXTERNAL_QUERY_GAS_LIMIT = 8000;
+
+    mapping (bytes32 => bool) public creditOrderCancelled;
+
+    event LogCreditOrderCancelled(
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash
+    );
+
+    event LogCreditOrderFilled(
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash,
+        bytes32 indexed _agreementId
+    );
+
+    event LogError(
+        uint8 indexed _errorId,
+        address indexed _creditor,
+        bytes32 indexed _creditorCommitmentHash
+    );
+
+    struct CreditorCommitment {
+        address creditor;
+        address repaymentVersion;
+        uint creditorFee;
+        address underwriter;
+        uint underwriterRiskRating;
+        address termsContract;
+        bytes32 termsContractParameters;
+        uint commitmentExpirationTimestampInSec;
+        uint salt;
+        bytes32 hash;
+    }
+
+    function CreditorProxy(address _tokenTransferProxy)
+        public
+    {
+        TOKEN_TRANSFER_PROXY = _tokenTransferProxy;
+    }
+
+    /**
+     * Allows contract owner to set the currently used debt token contract.
+     * Function exists to maximize upgradeability of individual modules
+     * in the entire system.
+     */
+    function setDebtToken(address debtTokenAddress)
+        public
+        onlyOwner
+    {
+        debtToken = DebtToken(debtTokenAddress);
+    }
+
+    /**
+     * Allows contract owner to set the currently used debt kernel contract.
+     * Function exists to maximize upgradeability of individual modules
+     * in the entire system.
+     */
+    function setDebtKernel(address debtKernelAddress)
+        public
+        onlyOwner
+    {
+        debtKernel = DebtKernel(debtKernelAddress);
+    }
+
+
+    /*
+     * Submit debt order to DebtKernel if it is consensual with creditor's request
+     * Creditor signature in arguments is external creditor and not relevant to Debt Kernel
+     */
+    function fillCreditOrder(
+        address creditor,
+        address[6] orderAddresses, // repayment-router, debtor, uw, tc, p-token, relayer
+        uint[8] orderValues, // rr, salt, pa, uwFee, rFee, cFee, dFee, expTime
+        bytes32[1] orderBytes32, // tcParams
+        uint8[3] signaturesV, // debtV, credV, uwV
+        bytes32[3] signaturesR,
+        bytes32[3] signaturesS
+    )
+    public
+    whenNotPaused
+    returns (bytes32 _agreementId)
+    {
+        uint principalAmount = orderValues[2];
+        address principalToken = orderAddresses[4];
+
+        CreditorCommitment memory creditorCommitment = getCreditorCommitment(
+            creditor, orderAddresses, orderValues, orderBytes32
+        );
+
+        uint totalCreditorPayment = principalAmount.add(creditorCommitment.creditorFee);
+
+        if (!assertNoReplay(creditorCommitment.hash)) { 
```

Comment:
I think `assertNoReplay` can be more general; i.e., something like `assertOrderNotFilled`

Reply:
After separating the way we record cancelled and filled orders (per one of your later comments), I'm inclined to just check storage directly in the `if` statement & eliminate this helper function all together..